### PR TITLE
ref: use macos arm builders to speed up library build

### PR DESCRIPTION
**Pull request by asottile-sentry**

### CODE REVIEW

File: .github/workflows/build_library.yml [hunk 1]
```diff
@@ -47,12 +47,12 @@ jobs:
       matrix:
         include:
           - target: x86_64-apple-darwin
-            py-platform: macosx-10_15_x86_64
+            py-platform: macosx-14_0_x86_64
```
Comment (jan-auer) [inline .github/workflows/build_library.yml:50]: Do we require >= macOS 14 now for all of sentry development? If not, I'd at least keep the intel build where it was. 

Technically, there shouldn't be any problems, I noticed that even built on newer versions binaries require a lower version number of the system libraries, but still...

**Review by jan-auer — COMMENTED** [review]
We can also update the workflows in `ci.yml` and `build_binary.yml`

File: .github/workflows/build_library.yml (hunk 1 above)
Comment (asottile-sentry) [reply .github/workflows/build_library.yml:50]: the value here ~generally needs to match the platform it is built from -- it's already wrong on master :)

but yes we already require macos>=14

**Review by jan-auer — APPROVED** [review]
:shipit:

**Pull request merged**